src/content/preprocessed-profile-versioning.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * This file deals with old versions of the "preprocessed" profile format,
+ * i.e. the format that perf.html uses internally. Profiles in this format
+ * can be saved out to files or uploaded to the profile store server, and we
+ * want to be able to display profiles that were saved at any point in the
+ * past, regardless of their version. So this file upgrades old profiles to
+ * the current format.
+ */
+
+import { sortDataTable } from './data-table-utils';
+
+export const CURRENT_VERSION = 1; // The current version of the 'preprocessed profile' format.
+
+// Preprocessed profiles before version 1 did not have a profile.meta.preprocessedProfileVersion
+// field. Treat those as version zero.
+const UNANNOTATED_VERSION = 0;
+
+export function isPreprocessedProfile(profile) {
+  // If this profile has a .meta.preprocessedProfileVersion field,
+  // then it is definitely a preprocessed profile.
+  if ('meta' in profile && 'preprocessedProfileVersion' in profile.meta) {
+    return true;
+  }
+
+  // This could also be a pre-version 1 profile.
+  return 'threads' in profile &&
+    profile.threads.length >= 1 &&
+    'stringArray' in profile.threads[0];
+}
+
+/**
+ * Upgrades the supplied profile to the current version, by mutating |profile|.
+ * Throws an exception if the profile is too new.
+ * @param {object} profile The "serialized" form of a preprocessed profile,
+ *                         i.e. stringArray instead of stringTable.
+ */
+export function upgradePreprocessedProfileToCurrentVersion(profile) {
+  const profileVersion = profile.meta.preprocessedProfileVersion ||
+    UNANNOTATED_VERSION;
+  if (profileVersion === CURRENT_VERSION) {
+    return;
+  }
+
+  if (profileVersion > CURRENT_VERSION) {
+    throw new Error(
+      `Unable to parse a preprocessed profile of version ${profileVersion} - are you running an outdated version of perf.html? ` +
+        `The most recent version understood by this version of perf.html is version ${CURRENT_VERSION}.\n` +
+        'You can try refreshing this page in case perf.html has updated in the meantime.'
+    );
+  }
+
+  // Convert to CURRENT_VERSION, one step at a time.
+  for (
+    let destVersion = profileVersion + 1;
+    destVersion <= CURRENT_VERSION;
+    destVersion++
+  ) {
+    if (destVersion in _upgraders) {
+      _upgraders[destVersion](profile);
+    }
+  }
+
+  profile.meta.preprocessedProfileVersion = CURRENT_VERSION;
+}
+
+// _upgraders[i] converts from version i - 1 to version i.
+// Every "upgrader" takes the profile as its single argument and mutates it.
+const _upgraders = {
+  1: profile => {
```

Comment:
It's a bit of a lie, since object property names are implicitly stringified anyway... and now that I'm flow-checking this file, it's complaining about it, so I'll make them string keys.